src/optimizer/OptimizerUtils.cpp

```diff
@@ -463,14 +482,399 @@ Value OptimizerUtils::normalizeValue(const meta::cpp2::ColumnDef& col, const Val
-                        << "for index.";
+            DLOG(FATAL) << "Not supported value type " << type << "for index.";
             return Value::kNullBadType;
         }
     }
     DLOG(FATAL) << "Unknown value type " << static_cast<int>(type);
-    return Value::kNullBadType;;
+    return Value::kNullBadType;
+}
+
+Status OptimizerUtils::boundValue(Expression::Kind kind,
+                                  const Value& val,
+                                  const meta::cpp2::ColumnDef& col,
+                                  Value& begin,
+                                  Value& end) {
+    if (val.type() != graph::SchemaUtil::propTypeToValueType(col.type.type)) {
+        return Status::SemanticError("Data type error of field : %s", col.get_name().c_str());
+    }
+    switch (kind) {
+        case Expression::Kind::kRelLE: {
+            // if c1 <= int(5) , the range pair should be (min, 6)
+            // if c1 < int(5), the range pair should be (min, 5)
+            auto v = OptimizerUtils::boundValue(col, BoundValueOperator::GREATER_THAN, val);
+            if (v == Value::kNullBadType) {
+                LOG(ERROR) << "Get bound value error. field : " << col.get_name();
+                return Status::Error("Get bound value error. field : %s", col.get_name().c_str());
+            }
+            // where c <= 1 and c <= 2 , 1 should be valid.
+            if (end.empty()) {
+                end = v;
+            } else {
+                end = v < end ? v : end;
+            }
+            break;
+        }
+        case Expression::Kind::kRelGE: {
+            // where c >= 1 and c >= 2 , 2 should be valid.
+            if (begin.empty()) {
+                begin = val;
+            } else {
+                begin = val < begin ? begin : val;
+            }
+            break;
+        }
+        case Expression::Kind::kRelLT: {
+            // c < 5 and c < 6 , 5 should be valid.
+            if (end.empty()) {
+                end = val;
+            } else {
+                end = val < end ? val : end;
+            }
+            break;
+        }
+        case Expression::Kind::kRelGT: {
+            // if c >= 5, the range pair should be (5, max)
+            // if c > 5, the range pair should be (6, max)
+            auto v = OptimizerUtils::boundValue(col, BoundValueOperator::GREATER_THAN, val);
+            if (v == Value::kNullBadType) {
+                LOG(ERROR) << "Get bound value error. field : " << col.get_name();
+                return Status::Error("Get bound value error. field : %s", col.get_name().c_str());
+            }
+            // where c > 1 and c > 2 , 2 should be valid.
+            if (begin.empty()) {
+                begin = v;
+            } else {
+                begin = v < begin ? begin : v;
+            }
+            break;
+        }
+        default: {
+            // TODO(yee): Semantic error
+            return Status::Error("Invalid expression kind.");
```

Comment:
Why don't change it to SemanticError directly.

Reply:
It's in optimizer module. It's better to do some checks in validator.